### PR TITLE
add SampleYAML function

### PR DESCRIPTION
**Pull request by rogpeppe**
This will enable us to generate environment provider boilerplate
YAML files.

### CODE REVIEW

File: sample.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+		if previousText {
+			fmt.Fprintf(w, "%s\n", strings.TrimRightFunc(indent, unicode.IsSpace))
+		}
+		previousText = true
+	}
+
+	descr := strings.TrimSpace(f.Description)
+	if descr != "" {
+		section()
+		if !strings.HasSuffix(descr, ".") {
```
Comment (mhilton) [inline sample.go:72]: Are you sure you want to do this? Sentences might end with a ! or a ? (although that might be unusual for a description)

File: sample.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+			fmt.Fprintf(w, "# %s:", f.name)
+			indentVal(w, val, indentStr+"# ")
+		}
+	}
+	return nil
+}
+
+const textWidth = 80
+
+func writeSampleDescription(w io.Writer, f Attr, indent string) {
```
Comment (mhilton) [inline sample.go:57]: It might be useful to have a doc comment here.

File: sample.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+		attrText = "immutable"
+	}
+	if attrText != "" {
+		section()
+		fmt.Fprintf(w, "%sThis attribute is %s.\n", indent, attrText)
+	}
+	section()
+}
+
+func emptyLine(w io.Writer, indent string) {
```
Comment (mhilton) [inline sample.go:104]: A doc comment here would also be good please.

File: sample.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+				val = sampleValue(f.Type)
+			}
+			fmt.Fprintf(w, "# %s:", f.name)
+			indentVal(w, val, indentStr+"# ")
+		}
+	}
+	return nil
+}
+
+const textWidth = 80
```
Comment (urosj) [inline sample.go:55]: I know it's just a bit of paperwork to deal with indenting, however, i wonder if this is really required? What if we rather leave this to a user?

File: sample.go (hunk 1 above)
Comment (urosj) [inline sample.go:72]: +1 on this. Do as little formatting as possible.

File: sample.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+func (f fieldsByGroup) Len() int {
+	return len(f)
+}
+
+func (f fieldsByGroup) Swap(i, j int) {
+	f[i], f[j] = f[j], f[i]
+}
+
+func (f fieldsByGroup) Less(i0, i1 int) bool {
```
Comment (mhilton) [inline sample.go:139]: I think I'd like to see the same parameters for Less as Swap. I don't mind which ones you choose.

File: sample.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+		panic(fmt.Errorf("cannot marshal YAML", err))
+	}
+	if len(data) == 0 {
+		panic("YAML cannot marshal to empty string")
+	}
+	indent := []byte(indentStr + "  ")
+	if canUseSameLine(x) {
+		w.Write(space)
+	} else {
+		w.Write(nl)
```
Comment (mhilton) [inline sample.go:167]: previously you've used fmt.Fprintf(w, "\n") to do the same thing. I'm not following why this one should be different.

File: sample_test.go [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+		"baz": false,
+		"attrs": map[string]string{
+			"arble": "bletch",
+			"hello": "goodbye",
+		},
+	},
+	fields: environschema.Fields{
+		"foo": {
+			Type:        environschema.Tstring,
+			Description: "foo is a string.",
```
Comment (urosj) [inline sample_test.go:38]: Test with long description and sentences ending with ? and ! instead of .

File: sample.go [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+		w.Write(indent)
+	}
+	data = bytes.TrimSuffix(data, nl)
+	lines := bytes.Split(data, nl)
+	for i, line := range lines {
```
Comment (mhilton) [inline sample.go:172]: any particular reason not to merge this and the above line together?
`for i, line := bytes.Split(data, nl)` No particular reason to do so either.

File: sample_test.go [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+		|# c: ""
+		|
+		|# This attribute is immutable.
+		|#
+		|# d: ""
+		|
+		|# This attribute is immutable and considered secret.
+		|#
+		|# e: ""
+	`,
```
Comment (urosj) [inline sample_test.go:285]: What about nested attributes? 
level1:
\ \ level2:
\ \ \ \ level3:

**Comment by urosj** [comment]
Looks ok. I have some reservations about how output is generated (vs if user defines it), but as it is under our control, it shouldn't be a problem.

**Comment by mhilton** [comment]
:+1: with a couple of minor thoughts.

File: sample.go (hunk 4 above)
Comment (rogpeppe) [inline sample.go:55]: I _think_ it's worth doing, as it means we don't need to bother too much
about formatting and line lengths in the descriptions - they'll all be
formatted consistently (like Go doc comments)

File: sample.go (hunk 1 above)
Comment (rogpeppe) [inline sample.go:72]: Yeah, I agree (particularly since something might easily end up with an exclamation mark). Done.

File: sample.go (hunk 2 above)
Comment (rogpeppe) [inline sample.go:57]: done

File: sample.go (hunk 5 above)
Comment (rogpeppe) [inline sample.go:139]: done

File: sample.go (hunk 6 above)
Comment (rogpeppe) [inline sample.go:167]: i've made them consistent

File: sample.go (hunk 8 above)
Comment (rogpeppe) [inline sample.go:172]: I think it's quite nice to make it clear we have a slice of lines.

File: sample_test.go (hunk 9 above)
Comment (rogpeppe) [inline sample_test.go:285]: Not possible with the current environment schema.

**Comment by urosj** [comment]
:+1:

File: sample_test.go (hunk 7 above)
Comment (rogpeppe) [inline sample_test.go:38]: I've removed the full-stop-adding logic.

**Pull request merged**